### DISCUSSION TITLE
Why mismatch?

### DISCUSSION THREAD

**Discussion by vega0** [antlr/antlr4#4263] [Q&A]
Why tipBulevo: input 'boolean' doesn't match token value?? Periodically stucks on it when complicating .

https://pastebin.com/g6TDXyGv

**Comment by jimidle** [comment]
I think you need to take a step back and u Der stand how the rules work.
Read through sone of the tutorials. Your mistakes will make sense when you
understand how it works.

**Reply by vega0** [reply]
ehm, what exactly i'am doing wrong?

**Reply by kaby76** [reply]
> ehm, what exactly i'am doing wrong?

All you provide is a link to a grammar, YapComplete. We don't know the start symbol for the grammar. You should have in the YapComplete.g4 file a parser rule of the form `start_symbol : .............. EOF;`or tell us "The start symbol is 'probelniy'." or whatever the parser rule name is. For Bison grammars, [the start symbol is declared](https://www.gnu.org/software/bison/manual/html_node/Start-Decl.html). Antlr has no such constraint, but we need it to know why the input isn't parsing. We also do not know the input string. Is it just `boolean`?

**Comment by jimidle** [comment]
You are trying to write what you think it should do, and not what it does.
Take a step back and learn a little more. Start with the very simplest of
grammars to see how the lexer tokenizes everything, then the parser runs.
Only one lexer rule can win, and it isn’t influenced by the parser.

Also, put no ‘literals’ in the parser grammar.

**Reply by vega0** [reply]
The token rules in parser rules that in one parser rule . What is wrong?

Token lexer rule determines the form of word that uses in parser rules . 


```
OKONCHANIYA: ('ый' | 'ая' | 'ое');
TOKEN_VIDIMOSTI_PUBLICHNIY_1: 'публичн' OKONCHANIYA;
TOKEN_VIDIMOSTI_PRIVATNIY_1: 'приватн' OKONCHANIYA;
TOKEN_VIDIMOSTI_PUBLIC_ENG: 'public';
TOKEN_VIDIMOSTI_PRIVATE_ENG: 'private';
TOKEN_OUBLICHNIY_POLN: TOKEN_VIDIMOSTI_PUBLICHNIY_1 |  TOKEN_VIDIMOSTI_PUBLIC_ENG;
TOKEN_PRIVATNIY_POLN: TOKEN_VIDIMOSTI_PRIVATNIY_1 | TOKEN_VIDIMOSTI_PRIVATE_ENG;

// ФАЙЛ ОБЪХЕКТА TokenTipPeremennoy
TOKEN_TIPA_CHISLA: ('число' ('вой' | 'вое')? ' значение'?) | 'number';
TOKEN_TIPA_STROKA: ('строк' ('а' | 'о' ( 'в' OKONCHANIYA ) )) | 'string';
TOKEN_TIPA_FUNCIYA: ('функци' ('я' | 'ональный') | 'метод') | 'function' | 'method';
TOKEN_TIPA_AVTOMAT: ('авто' 'матический'?) | 'auto';
TOKEN_TIPA_BULEVO_ZNACHENIYE: ('булев' (('о' | 'ое') | OKONCHANIYA) ' значение'?) | 'boolean';
```

**Reply by jimidle** [reply]
There is a lot wrong with this. You are trying to make the lexer do the parsing.

You rule: `TOKEN_VIDIMOSTI_PUBLIC_ENG` matches 'public', but then your rule: `TOKEN_OUBLICHNIY_POLN` tries to match it again. If those two tokens are at different places in the parse, then they need to be parser rules, not lexer rules.

```antlr
TOKEN_VIDIMOSTI_PUBLIC_ENG: 'public';
TOKEN_VIDIMOSTI_PRIVATE_ENG: 'private';

pubpriv: TOKEN_VIDIMOSTI_PUBLIC_ENG | TOKEN_VIDIMOSTI_PRIVATE_ENG ;
```

If `OKONCHANIYA` does not exist on its own as a token, then make it a `fragment`.

Don't make your lexer rules look for the alternatives, like you are doing with `TOKEN_TIPA_FUNCIYA`, that is the job of the parser, not the lexer. make `TOKEN_TIPA_FUNCIYA` separate tokens and say where they can be seen in the parser.

Remember, that the parser does not tell the lexer what to do. The lexer makes ALL the tokens first, and THEN the parser consumes them. It does not matter how you specify the parser rules. I strongly recommend reading some tutorials, or maybe buying the book.  

I hope that this gets you started down the correct path. What you have now cannot work at all.

**Reply by vega0** [reply]
~ | is only parser rules?

**Reply by vega0** [reply]
but its work separately.

**Reply by vega0** [reply]
```
grammar StrokaFailovyPut;
import Oglavleniye;

fragmentputi: FRAGMENTPUTI rashireniyeFaila?;
rashireniyeFaila: Tochka fragmentputi;

failovyPut: SOBAKA DELITEL_PUTI? fragmentputi (DELITEL_PUTI fragmentputi)*;

SOBAKA: '@';
DELITEL_PUTI: '/';
EKRANIROVANNIY_DELITEL_PUTI: '\\ ';
FRAGMENTPUTI: (EKRANIROVANNIY_DELITEL_PUTI | ~(' ' | '/' | '.'))+;
```

why i can't use token defined in a set **~(' ' | '/' | '.')**? It seems duplicates.

**Reply by jimidle** [reply]
What are you trying to do here?

```antlr
fragmentputi: FRAGMENTPUTI rashireniyeFaila?;
rashireniyeFaila: Tochka fragmentputi;
```
Your FRAGMENTPUTI makes no sense.

Take a step back here as I think this may be an XY question:

What is it that you are trying to do?
Is this for your company or for a private project?

**Reply by vega0** [reply]
form of filepath. @/hello/world/path\ object/file.exe

**Reply by vega0** [reply]
not work correctly as token rule , needto do something like 

```
fragmentputi1: FRAGMENTPUTI;
fragmentputi: fragmentputi1 rashireniyeFaila?;
```

?? If i put all lexems into parser rule its work fine but i can't use tokens and i cause of token symbols duplicates.

**Reply by vega0** [reply]
https://i.imgur.com/oz41FEN.png it's work, but not after join into one file .

**Reply by vega0** [reply]
https://i.imgur.com/xUZw9J7.png because TOKEN MISMATCH, why .... and what.

**Comment by jimidle** [comment]
Because you are not making lexer tokens out of them, so ',' in the parser
is not the same ',' as it is in the lexer. I did say not use `literals` in
the parser. You want:

COMMA: ',' ;
DOT: '.' ;

In your lexer, and then use those tokens in your parser. Do not have any
literals in your parser:

parserRule:  expression (COMMA expression)* ;

And so on.

**Reply by vega0** [reply]
How to understand it?

```
warning(125): YapComplete.g4:80:33: implicit definition of token OKONCHANIYA in parser
YapComplete::modificatorKonstanty:1:0: mismatched input 'konstant' expecting {'константн', 'konstant', 'constant'}
warning(125): YapComplete.g4:33:14: implicit definition of token FRAGMENTPUTI in parser
warning(125): YapComplete.g4:80:33: implicit definition of token OKONCHANIYA in parser
YapComplete::modificatorKonstanty:1:0: mismatched input 'constant' expecting {'константн', 'konstant', 'constant'}
```

https://i.imgur.com/ksWKHjs.png

obvious must match but its mismatch ?

**Reply by vega0** [reply]
`YapComplete::failovyPut:1:0: mismatched input '@' expecting '@'`

**Comment by jimidle** [comment]
I have already told you.

The lexer returns a token for '@' but you have '@' in your parser. So they
are two different tokens, despite the text being the same. You need to
define EVERYTHING in the lexer if you want a separate lexer and parser:

```antlr
myrule: rule1 AT rule2 ;

AT: '@' ;
```

The implicit definition warnings are also telling you that there is no
lexer rule OKONCHANIYA in the lexer, so you have referred to it in the
parser, but there is no such token. Get your lexer working first, then you
can build the parser.

I implore you to read the tutorials. This is ANTLR 101 stuff that you need
to know before trying to write a grammar of any complexity. You might find
it easier not to separate the lexer and parser grammars, but it is
essentially the same thing.

As it stands, your grammars just won't work. Maybe attach them to this
conversation. But we don't really have time to give you free individual
training over GitHub I am afraid.

**Reply by vega0** [reply]
i can'tcreate set of tokens. :(((

**Reply by vega0** [reply]
https://www.youtube.com/watch?v=SABlWeviUUU

**Reply by kaby76** [reply]
1) You must fix warnings from the Antlr tool in your grammar, e.g., "implicit definition of ... in parser". Your grammar cannot work otherwise.

2) You keep changing the grammar from one comment to the next. I cannot tell what you are working on at all. You now have an "import". 

3) Stop using Eclipse. Write a driver program and, set up a script to run the Antlr4 tool, compile and link your code. Instead of posting .rar files, pictures, and movies, please check in the code in a Github repo so we can factually *test* the code on the exact version of your code. If you cannot write a driver, give me the grammar files, tell me the start rule, the target (CSharp, Java, JavaScript, Go, Dart2, PHP, Python3, ...) and I will generate a driver and build scripts for you.

4) In your driver code, print out the tokens recognized. Don't even try to call the parser. Your lexer rules are completely messed up.

5) As far as I can tell, you still do not have an EOF-terminated start rule. You *must* have exactly one EOF-terminated start rule.

**Reply by vega0** [reply]
https://gitflic.ru/project/nightmare0/rossijskij-yap

**Reply by kaby76** [reply]
OK, got it. I will step through your code and try to clean it up.

1) I copied the .g4's to my own github.com repo [here](https://github.com/kaby76/antlr4-4263). https://gitflic.ru/ does not offer permalinks to lines of code--too primitive to use. That is why I made a copy to https://github.com/kaby76/antlr4-4263. Github provides a much better git experience.

2) I changed all the single-line imports to multiline imports. For example `import Peremenniye, StrokaFailovyPut, Oglavleniye;` to `import Peremenniye;
import StrokaFailovyPut;
import Oglavleniye;`, with each import on a separate line for readability. I mentioned before: you should stop using `import`!! It really makes the grammar extremely confusing. And, I don't know whether Antlr *orders* imports. Nobody can remember, and in order to know, you have to constantly test and check what Antlr does. So, just make one grammar file. After you get a combined grammar working, you can then split it up to your heart's content.

3) There is only one .g4 with an EOF-terminated start rule, [here](https://github.com/kaby76/antlr4-4263/blob/5a2a7db024a676804bc7169ecdb7340d9222fb72/InstrukciyaImportirovat.g4#L14). Therefore, I will assume that is the top-level grammar and `importirovatInstrukciyaChereda` is the start symbol.

4) I wrote a [desc.xml](https://github.com/kaby76/antlr4-4263/blob/5a2a7db024a676804bc7169ecdb7340d9222fb72/desc.xml) file, then generated a driver from the source using [trgen](https://github.com/kaby76/Domemtech.Trash/tree/main/src/trgen). The output from the build of the generated driver is:

```
05/11-13:46:03 ~/l/x/z/fff
$ ls
buildfail.py                InstrykciyaFunkciya.g4      Oglavleniye.g4       Stroki.g4                         TokenTipPeremennoy.g4
Chisla2.g4                  InstrykciyaKlass.g4         Peremenniye.g4       TokenModificatorVidimosti.g4      Virajeniya.g4
desc.xml                    InstrykciyaVizovObiekta.g4  README.md            TokenModifikatorKonstantnosti.g4  Yap.g4
InstrukciyaImportirovat.g4  LICENSE                     StrokaFailovyPut.g4  TokenModifikatory.g4              YapComplete.g4
05/11-13:46:04 ~/l/x/z/fff
$ trgen -t CSharp
C:\msys64\home\Kenne\l\x\z\fff
CSharp  InstrukciyaImportirovat.g4 success 0.0502904
Rendering template file from CSharp/build.ps1 to Generated-CSharp/build.ps1
Rendering template file from CSharp/build.sh to Generated-CSharp/build.sh
Rendering template file from CSharp/clean.ps1 to Generated-CSharp/clean.ps1
Rendering template file from CSharp/clean.sh to Generated-CSharp/clean.sh
Rendering template file from CSharp/Encodings.cs to Generated-CSharp/Encodings.cs
Rendering template file from CSharp/ErrorListener.cs to Generated-CSharp/ErrorListener.cs
Rendering template file from CSharp/makefile to Generated-CSharp/makefile
Rendering template file from CSharp/perf.sh to Generated-CSharp/perf.sh
Rendering template file from CSharp/run.ps1 to Generated-CSharp/run.ps1
Rendering template file from CSharp/run.sh to Generated-CSharp/run.sh
Rendering template file from CSharp/Test.cs to Generated-CSharp/Test.cs
Rendering template file from CSharp/Test.csproj to Generated-CSharp/Test.csproj
Rendering template file from CSharp/test.ps1 to Generated-CSharp/test.ps1
Rendering template file from CSharp/test.sh to Generated-CSharp/test.sh
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/YapComplete.g4 to Generated-CSharp/YapComplete.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/Yap.g4 to Generated-CSharp/Yap.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/Virajeniya.g4 to Generated-CSharp/Virajeniya.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/TokenTipPeremennoy.g4 to Generated-CSharp/TokenTipPeremennoy.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/TokenModifikatory.g4 to Generated-CSharp/TokenModifikatory.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/TokenModifikatorKonstantnosti.g4 to Generated-CSharp/TokenModifikatorKonstantnosti.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/TokenModificatorVidimosti.g4 to Generated-CSharp/TokenModificatorVidimosti.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/Stroki.g4 to Generated-CSharp/Stroki.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/StrokaFailovyPut.g4 to Generated-CSharp/StrokaFailovyPut.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/README.md to Generated-CSharp/README.md
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/Peremenniye.g4 to Generated-CSharp/Peremenniye.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/Oglavleniye.g4 to Generated-CSharp/Oglavleniye.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/LICENSE to Generated-CSharp/LICENSE
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/InstrykciyaVizovObiekta.g4 to Generated-CSharp/InstrykciyaVizovObiekta.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/InstrykciyaKlass.g4 to Generated-CSharp/InstrykciyaKlass.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/InstrykciyaFunkciya.g4 to Generated-CSharp/InstrykciyaFunkciya.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/InstrukciyaImportirovat.g4 to Generated-CSharp/InstrukciyaImportirovat.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/desc.xml to Generated-CSharp/desc.xml
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/Chisla2.g4 to Generated-CSharp/Chisla2.g4
Copying source file from C:/msys64/home/Kenne/l/x/z/fff/buildfail.py to Generated-CSharp/buildfail.py
05/11-13:46:14 ~/l/x/z/fff
$ cd Generated-CSharp/
05/11-13:46:17 ~/l/x/z/fff/Generated-CSharp
$ make
bash build.sh
  Determining projects to restore...
  Restored C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj (in 328 ms).
MSBuild version 17.6.0-preview-23179-01+dd5d9f746 for .NET
  Determining projects to restore...
  All projects are up-to-date for restore.
C:\Program Files\dotnet\sdk\7.0.300-preview.23179.2\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(
287,5): message NETSDK1057: You are using a preview version of .NET. See: https://aka.ms/dotnet-support-policy [C:\msys64\home\Kenne
\l\x\z\fff\Generated-CSharp\Test.csproj]
Peremenniye.g4(23,66): error ANT02: error(56): Peremenniye.g4:23:65: reference to undefined rule: DesyaterichniyeChisla [C:\msys64\h
ome\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
InstrukciyaImportirovat.g4(7,62): error ANT02: error(56): InstrukciyaImportirovat.g4:7:61: reference to undefined rule: failovyPut [
C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
C:\Users\Kenne\.nuget\packages\antlr4buildtasks\12.2.0\build\Antlr4BuildTasks.targets(138,9): error ANT02: The Antlr4 tool failed. [
C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]

Build FAILED.

Peremenniye.g4(23,66): error ANT02: error(56): Peremenniye.g4:23:65: reference to undefined rule: DesyaterichniyeChisla [C:\msys64\h
ome\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
InstrukciyaImportirovat.g4(7,62): error ANT02: error(56): InstrukciyaImportirovat.g4:7:61: reference to undefined rule: failovyPut [
C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
C:\Users\Kenne\.nuget\packages\antlr4buildtasks\12.2.0\build\Antlr4BuildTasks.targets(138,9): error ANT02: The Antlr4 tool failed. [
C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
    0 Warning(s)
    3 Error(s)

Time Elapsed 00:00:02.49
mingw32-make: *** [makefile:3: build] Error 1
05/11-13:46:23 ~/l/x/z/fff/Generated-CSharp
```

There is no definition of `DesyaterichniyeChisla` (used [here](https://github.com/kaby76/antlr4-4263/blob/5a2a7db024a676804bc7169ecdb7340d9222fb72/Peremenniye.g4#L23) and [here](https://github.com/kaby76/antlr4-4263/blob/5a2a7db024a676804bc7169ecdb7340d9222fb72/YapComplete.g4#L156)).

***The grammar cannot work without defining all symbols.***

**Reply by vega0** [reply]
fixed all renames. https://gitflic.ru/project/nightmare0/rossijskij-yap/file?branch=master

**Reply by vega0** [reply]
yes, its confusing, i've used some single char tokens in parser rules and it gives error in assembled state (through buildfail.py
)

**Reply by kaby76** [reply]
OK, got that. 

Looks like splitting up the imports into multiple lines causes a build error with the Antlr4 tool (see below). Using your grammar files exactly as is gives only warnings, but these need to be corrected! It seems like there might be a problem with the Antlr4 tool. So, I highly recommend you combine the grammars to avoid any scoping issues. In the meanwhile, I'll investigate separate-line imports vs combined-line import. Something doesn't feel right.
```
InstrukciyaImportirovat.g4(16,24): warning ANT01: warning(180): InstrukciyaImportirovat.g4:16:23: chars 0-? used multiple times in s
et [\u10000-\u10FFFF] [C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
InstrukciyaImportirovat.g4(16,24): warning ANT01: warning(180): InstrukciyaImportirovat.g4:16:23: chars F used multiple times in set
 [\u10000-\u10FFFF] [C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
InstrukciyaImportirovat.g4(16,24): warning ANT01: warning(180): InstrukciyaImportirovat.g4:16:23: chars F used multiple times in set
 [\u10000-\u10FFFF] [C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
InstrukciyaImportirovat.g4(7,62): error ANT02: error(56): InstrukciyaImportirovat.g4:7:61: reference to undefined rule: failovyPut [
C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
C:\Users\Kenne\.nuget\packages\antlr4buildtasks\12.2.0\build\Antlr4BuildTasks.targets(138,9): error ANT02: The Antlr4 tool failed. [
C:\msys64\home\Kenne\l\x\z\fff\Generated-CSharp\Test.csproj]
    3 Warning(s)
    2 Error(s)
```

1)

**Reply by vega0** [reply]
can you tell me build instructions or script to test it itself and why this error not happens in eclipse?

**Reply by vega0** [reply]
why failovyPut in base file is work correctly but after importing it into anotrer place its throws error.

**Reply by kaby76** [reply]
I've updated https://github.com/kaby76/antlr4-4263 to include your original set of grammars in [t1](https://github.com/kaby76/antlr4-4263/tree/0e93645e618a81341e4f219a90e76f5c769f7f0b/t1) and the modified set of grammars (from single-line imports "import a, b, c, d;" to sequential-line imports "import a; import b; import c; import d;") in [t2](https://github.com/kaby76/antlr4-4263/tree/0e93645e618a81341e4f219a90e76f5c769f7f0b/t2). [t1](https://github.com/kaby76/antlr4-4263/tree/0e93645e618a81341e4f219a90e76f5c769f7f0b/t1) builds (via `dotnet build`), [t2](https://github.com/kaby76/antlr4-4263/tree/0e93645e618a81341e4f219a90e76f5c769f7f0b/t2) does not.

This means that the scoping of the symbols is not flat. The description of `import` is in the [Documentation here](https://github.com/antlr/antlr4/blob/master/doc/grammars.md#grammar-imports).
> Rules in the “main grammar” override rules from imported grammars to implement inheritance.

There seems to be an issue with an import graph that is a [partial order](https://en.wikipedia.org/wiki/Partially_ordered_set#Partial_order). As I shown in the diagram [here](https://github.com/antlr/antlr4/discussions/4262#discussioncomment-5851800), grammar `StrokaFailovyPut`, which defines parse rule symbol `failovyPut`, is imported on two paths, one directly from grammar `InstrukciyaImportirovat -> StrokaFailovyPut`, and the other via path `InstrukciyaImportirovat -> Peremenniye -> Stroki -> StrokaFailovyPut`. Apparently, that causes problems with Antlr. It's likely you could refactor the grammar so that it's a [total order](https://en.wikipedia.org/wiki/Total_order) to fix this, but I recommend collecting all rules into a single file.

You really should not be partitioning your grammar over 16 .g4 files. It is nearly impossible to read, and for sure requires an IDE to keep track of which symbol is in which file. I have never seen a grammar like this in my 40+ years of programming. And, there is nothing in the 300+ grammars of [grammars-v4](https://github.com/antlr/grammars-v4) remotely similar.

To build, you will need to download Microsoft's NET SDK version 7. Afterward, from a Bash shell prompt, "dotnet build".

The Eclipse, Intellij, and lab.antlr.org IDEs all uses an "interpretive" Antlr. It is not identical to that from the Antlr tool. There are subtle differences between "interpretive" and "compiled" Antlrs.

**Reply by vega0** [reply]
when i automatic to join all into one file this doesn't work then too.

**Reply by kaby76** [reply]
> when i automatic to join all into one file this doesn't work then too.

What doesn't work? Please show your work.

YapComplete.g4, which still does not have a start rule, has a duplicate rule for '/'.
```
05/13-05:48:32 ~/l/x/z/rossijskij-yap
$ gitr
origin  https://gitflic.ru/project/nightmare0/rossijskij-yap.git (fetch)
origin  https://gitflic.ru/project/nightmare0/rossijskij-yap.git (push)
05/13-05:48:41 ~/l/x/z/rossijskij-yap
$ git pull origin
Already up to date.
05/13-05:48:51 ~/l/x/z/rossijskij-yap
$ git status .
On branch master
Your branch is up to date with 'origin/master'.

nothing to commit, working tree clean
05/13-05:49:06 ~/l/x/z/rossijskij-yap
$ trparse -l YapComplete.g4 | trxgrep " //lexerRuleBlock/lexerAltList/lexerAlt/lexerElements/lexerElement/lexerAtom/terminal/STRING_LITERAL[text() = '''/''']/ancestor::lexerRuleSpec" | trtext
CSharp 0 YapComplete.g4 success 0.0791502

DELITELPUTI: '/';


// ???? ???????? Virajeniya
DELENIYE: '/';
05/13-05:49:16 ~/l/x/z/rossijskij-yap
$ git show
commit 79d00bb4eef825083943b5a3f7e383cab8ae9fdc (HEAD -> master, origin/master, origin/HEAD)
...
```
YapComplete.g4 cannot work because you define two lexer rules with the same right-hand side: `DELITELPUTI` and `DELENIYE`. Antlr will allow you to do that but at least outputs a warning. But, you should treat it as an error because only the first, `DELITELPUTI` will ever be returned by the lexer. Any rule that derives `DELENIYE` will not work.

***Antlr lexers are not parser aware at all. Antlr lexers operate by two rules: (1) match and return the token with the longest string--regardless of the order of the rules; (2) if two rules match the same string, the first one "wins".***

You have to remove the rule `DELENIYE` and replace all occurrences of it in any rules, e.g., in YapComplete.g4:
```
virajeniye: virajeniye operaciya=(UMNOJENIYE | DELITELPUTI | STEPENN) virajeniye // умножение / деление
		  | virajeniye operaciya=(A_ПЛЮС | A_МИНУС) 	   		   virajeniye // сложение / вычитание
		  | LEVAYASKOBKA virajeniye PRAVAYASKOBKA
		  | ssilkaNaObiekt
		  | aПолныйВидЧисла;
```
Again, *what is the start rule for which grammar?* If you are referring to YapComplete.g4, add an EOF-terminated rule.

**Reply by vega0** [reply]
Rules can be as much as possible complicated to make so complicated parsers. Why its not should be complicated? And yes, the question, buildscript.

**Reply by vega0** [reply]
sometimes lost utf8 recognition. Cause ? 
https://i.imgur.com/rsGqjRT.png

**Reply by kaby76** [reply]
> https://i.imgur.com/rsGqjRT.png

"Cause?" The "ANTLR Console" pane says that you are using antlr-4.4-complete.jar. That jar is over 9 years old. We are now on version 4.12. Until you change that, expect problems.

**Comment by jimidle** [comment]
I want to back Ken up here. Just put e Erin one file , get it working, then
worry about splitting it up if there is sone overriding reason.

I temper dealing with this in v3 and C and it was complicated, then nobody
used it.

**Reply by kaby76** [reply]
That's a *strong* thumbs up. Definitely, combine all .g4's into one. This is ridiculous.

**Reply by vega0** [reply]
combined and it wont work as in splitted state.

not work correctly as token rule , needto do something like

```
fragmentputi1: FRAGMENTPUTI;
fragmentputi: fragmentputi1 rashireniyeFaila?;
?? If i put all lexems into parser rule its work fine but i can't use tokens and i cause of token symbols duplicates.

@vega0
Comment options
[vega0](https://github.com/vega0)
[4 hours ago](https://github.com/antlr/antlr4/discussions/4263#discussioncomment-5890551)
Author
https://i.imgur.com/oz41FEN.png it's work, but not after join into one file .

@vega0
Comment options
[vega0](https://github.com/vega0)
[4 hours ago](https://github.com/antlr/antlr4/discussions/4263#discussioncomment-5890553)
Author
https://i.imgur.com/xUZw9J7.png because TOKEN MISMATCH, why .... and what.
```
